File: src/harness_core/dispatch/pasteback_parser.py
"""PastebackSubmission schema and PastebackParser for human-pasted output."""

from __future__ import annotations

import hashlib
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

# ---------------------------------------------------------------------------
# Schema version
# ---------------------------------------------------------------------------

PASTEBACK_SUBMISSION_SCHEMA_VERSION = "pasteback_submission.v1"

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

MAX_OUTPUT_LENGTH: int = 100_000  # 100k chars
ESTIMATED_CHARS_PER_TOKEN: int = 4
DEFAULT_COST_PER_1K_TOKENS: float = 0.002


# ---------------------------------------------------------------------------
# Schema
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class PastebackSubmission:
    submission_id: str
    dispatch_id: str
    submitted_by: str
    model_used: str | None
    provider_used: str | None
    raw_output: str
    output_hash: str
    claimed_input_tokens: int | None
    claimed_output_tokens: int | None
    claimed_cost: float | None
    submitted_at: str
    schema_version: str = PASTEBACK_SUBMISSION_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "submission_id": self.submission_id,
            "dispatch_id": self.dispatch_id,
            "submitted_by": self.submitted_by,
            "model_used": self.model_used,
            "provider_used": self.provider_used,
            "raw_output": self.raw_output,
            "output_hash": self.output_hash,
            "claimed_input_tokens": self.claimed_input_tokens,
            "claimed_output_tokens": self.claimed_output_tokens,
            "claimed_cost": self.claimed_cost,
            "submitted_at": self.submitted_at,
        }


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------


class PastebackParser:
    """Validates and parses human-pasted output into PastebackSubmission."""

    def parse(
        self,
        dispatch_id: str,
        raw_output: str,
        submitted_by: str = "human",
        model_used: str | None = None,
        provider_used: str | None = None,
        claimed_input_tokens: int | None = None,
        claimed_output_tokens: int | None = None,
        claimed_cost: float | None = None,
    ) -> PastebackSubmission:
        validated_output = self._validate_output(raw_output)
        output_hash = self._hash_output(validated_output)

        return PastebackSubmission(
            submission_id=f"pb-{uuid.uuid4().hex[:12]}",
            dispatch_id=dispatch_id,
            submitted_by=submitted_by,
            model_used=model_used,
            provider_used=provider_used,
            raw_output=validated_output,
            output_hash=output_hash,
            claimed_input_tokens=claimed_input_tokens,
            claimed_output_tokens=claimed_output_tokens,
            claimed_cost=claimed_cost,
            submitted_at=datetime.now(timezone.utc).isoformat(),
        )

    def estimate_tokens(self, text: str) -> int:
        return max(1, len(text) // ESTIMATED_CHARS_PER_TOKEN)

    def estimate_cost(
        self, input_tokens: int, output_tokens: int, cost_per_1k: float = DEFAULT_COST_PER_1K_TOKENS
    ) -> float:
        return round((input_tokens + output_tokens) / 1000 * cost_per_1k, 6)

    def _validate_output(self, raw_output: str) -> str:
        if not raw_output or not raw_output.strip():
            raise ValueError("Pasteback output cannot be empty")
        output = raw_output.strip()
        if len(output) > MAX_OUTPUT_LENGTH:
            raise ValueError(
                f"Output exceeds max length ({len(output)} > {MAX_OUTPUT_LENGTH})"
            )
        return output

    def _hash_output(self, output: str) -> str:
        return hashlib.sha256(output.encode("utf-8")).hexdigest()[:16]

File: engine/src/dispatch/manual/pasteback_parser.rs
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use uuid::Uuid;

pub const PASTEBACK_SUBMISSION_SCHEMA_VERSION: &str = "pasteback_submission.v1";
pub const MAX_OUTPUT_LENGTH: usize = 100_000;
pub const ESTIMATED_CHARS_PER_TOKEN: usize = 4;
pub const DEFAULT_COST_PER_1K_TOKENS: f64 = 0.002;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct PastebackSubmission {
    pub schema_version: String,
    pub submission_id: String,
    pub dispatch_id: String,
    pub submitted_by: String,
    pub model_used: Option<String>,
    pub provider_used: Option<String>,
    pub raw_output: String,
    pub output_hash: String,
    pub claimed_input_tokens: Option<i64>,
    pub claimed_output_tokens: Option<i64>,
    pub claimed_cost: Option<f64>,
    pub submitted_at: String,
}

pub struct PastebackParser;

impl Default for PastebackParser {
    fn default() -> Self {
        Self
    }
}

impl PastebackParser {
    pub fn new() -> Self {
        Self
    }

    pub fn parse(
        &self,
        dispatch_id: &str,
        raw_output: &str,
        submitted_by: &str,
        model_used: Option<&str>,
        provider_used: Option<&str>,
        claimed_input_tokens: Option<i64>,
        claimed_output_tokens: Option<i64>,
        claimed_cost: Option<f64>,
    ) -> Result<PastebackSubmission, String> {
        let trimmed = raw_output.trim();
        if trimmed.is_empty() {
            return Err("Pasteback output cannot be empty".to_string());
        }
        if trimmed.len() > MAX_OUTPUT_LENGTH {
            return Err(format!(
                "Output exceeds max length ({} > {})",
                trimmed.len(),
                MAX_OUTPUT_LENGTH
            ));
        }
        let output_hash = {
            let mut h = Sha256::new();
            h.update(trimmed.as_bytes());
            hex::encode(&h.finalize()[..8])
        };
        Ok(PastebackSubmission {
            schema_version: PASTEBACK_SUBMISSION_SCHEMA_VERSION.to_string(),
            submission_id: format!("pb-{}", &Uuid::new_v4().to_string().replace('-', "")[..12]),
            dispatch_id: dispatch_id.to_string(),
            submitted_by: submitted_by.to_string(),
            model_used: model_used.map(|s| s.to_string()),
            provider_used: provider_used.map(|s| s.to_string()),
            raw_output: trimmed.to_string(),
            output_hash,
            claimed_input_tokens,
            claimed_output_tokens,
            claimed_cost,
            submitted_at: chrono::Utc::now().to_rfc3339(),
        })
    }

    pub fn estimate_tokens(&self, text: &str) -> i64 {
        std::cmp::max(1, (text.len() / ESTIMATED_CHARS_PER_TOKEN) as i64)
    }

    pub fn estimate_cost(&self, input_tokens: i64, output_tokens: i64, cost_per_1k: f64) -> f64 {
        ((input_tokens + output_tokens) as f64 / 1000.0 * cost_per_1k * 1_000_000.0).round()
            / 1_000_000.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_valid() {
        let p = PastebackParser::new();
        let s = p
            .parse("d1", "Hello world", "human", None, None, None, None, None)
            .unwrap();
        assert_eq!(s.dispatch_id, "d1");
        assert!(s.submission_id.starts_with("pb-"));
    }

    #[test]
    fn parse_empty_rejected() {
        assert!(PastebackParser::new()
            .parse("d1", "", "human", None, None, None, None, None)
            .is_err());
    }

    #[test]
    fn estimate_tokens() {
        assert_eq!(PastebackParser::new().estimate_tokens("hello"), 1);
    }

    #[test]
    fn estimate_cost() {
        let cost = PastebackParser::new().estimate_cost(1000, 500, 0.002);
        assert!((cost - 0.003).abs() < 0.0001);
    }

    #[test]
    fn to_value_roundtrip() {
        let p = PastebackParser::new();
        let s = p
            .parse(
                "d1",
                "out",
                "h",
                Some("g"),
                Some("o"),
                Some(1),
                Some(2),
                Some(0.1),
            )
            .unwrap();
        let v = serde_json::to_value(&s).unwrap();
        let d: PastebackSubmission = serde_json::from_value(v).unwrap();
        assert_eq!(d, s);
    }
}
